gnovm/pkg/gnolang/preprocess.go

```diff
@@ -1961,60 +1961,15 @@ func preprocess1(store Store, ctx BlockNode, n Node) Node {
 						convertIfConst(store, last, rx)
 					}
 
-					if len(n.Lhs) > len(n.Rhs) {
-						switch cx := n.Rhs[0].(type) {
-						case *CallExpr:
-							// Call case: a, b := x(...)
-							ift := evalStaticTypeOf(store, last, cx.Func)
-							cft := getGnoFuncTypeOf(store, ift)
-							for i, lx := range n.Lhs {
-								ln := lx.(*NameExpr).Name
-								rf := cft.Results[i]
-								// re-definition
-								last.Define(ln, anyValue(rf.Type))
-							}
-						case *TypeAssertExpr:
-							lhs0 := n.Lhs[0].(*NameExpr).Name
-							lhs1 := n.Lhs[1].(*NameExpr).Name
-							tt := evalStaticType(store, last, cx.Type)
-							// re-definitions
-							last.Define(lhs0, anyValue(tt))
-							last.Define(lhs1, anyValue(BoolType))
-						case *IndexExpr:
-							lhs0 := n.Lhs[0].(*NameExpr).Name
-							lhs1 := n.Lhs[1].(*NameExpr).Name
-
-							var mt *MapType
-							dt := evalStaticTypeOf(store, last, cx.X)
-							mt, ok := baseOf(dt).(*MapType)
-							if !ok {
-								panic(fmt.Sprintf("invalid index expression on %T", dt))
-							}
-							// re-definitions
-							last.Define(lhs0, anyValue(mt.Value))
-							last.Define(lhs1, anyValue(BoolType))
-						default:
-							panic("should not happen")
-						}
-					} else {
-						// General case: a, b := x, y
-						for i, lx := range n.Lhs {
-							ln := lx.(*NameExpr).Name
-							rx := n.Rhs[i]
-							rt := evalStaticTypeOf(store, last, rx)
-							// re-definition
-							if rt == nil {
-								// e.g. (interface{})(nil), becomes ConstExpr(undefined).
-								// last.Define(ln, undefined) complains, since redefinition.
-							} else {
-								last.Define(ln, anyValue(rt))
-							}
-							// if rhs is untyped
-							if isUntyped(rt) {
-								checkOrConvertType(store, last, &n.Rhs[i], nil, false)
-							}
-						}
+					numNames := len(n.Lhs)
+					sts, tvs, nameExprs := prepareTypeVals(n, numNames)
+
+					if numNames > len(n.Rhs) { // Special cases
+						specialParseTypeVals(sts, tvs, store, last, nameExprs, nil, n.Rhs[0])
+					} else { // General case
+						generalParseTypeVals(sts, tvs, store, last, false, nameExprs, nil, n.Rhs)
```

Comment:
maybe we should have just one function for this ?

Reply:
I'd like to do it but it turns out that we have to make if else in that function for Assign and ValueDecl. I don't have any solution to merge those both functions for now.

Reply:
The check we implemented on ValueDecl should also apply to the assignment part, as we currently lack that verification.

Reply:
> The check(if/else) we implemented on ValueDecl should also apply to the assignment part, as we currently lack that verification.

currently it's not lack, it's all in `AssertCompatible` function